Firestore/core/src/firebase/firestore/remote/buffered_writer.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_BUFFERED_WRITER_H
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_BUFFERED_WRITER_H
+
+#include <functional>
+#include <vector>
+
+#include <grpcpp/support/byte_buffer.h>
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+// GRPC doesn't allow issuing a write operation before the previous write has
+// finished. This class helps keep track of pending writes and whether there
+// currently is a write operation in progress; it doesn't communicate with GRPC
+// directly and expects the following to function correctly:
+// - the constructor should be given a function that actually issues the GRPC
+//   write operation;
+// - each time a write operation finishes, `OnSuccessfulWrite` must be called.
+// The invariant is that `BufferedWriter` will never invoke the writing function
+// if `OnSuccessfulWrite` hasn't been called since the previous invocation.
+//
+// The main methods are `Enqueue` and `OnSuccessfulWrite`. `Enqueue` will invoke
+// the writing function immediately if there is no operation in progress;
+// otherwise, the given `bytes` will be stored in the buffer.
+// `OnSuccessfulWrite` will invoke the writing function with the next write in
+// buffer, as long as the buffer is not empty (in FIFO order).
+
+// `BufferedWriter` can be `Start`ed and `Stop`ped, which is expected to reflect
+// the state of the GRPC call. When `BufferedWriter` is not started, writes can
+// be enqueued, but the writing function will not be invoked. Once the
+// `BufferedWriter` is started, it will immediately invoke the writing function
+// if the buffer is non-empty. When the `BufferedWriter` is first created, it's
+// in the stopped state. `BufferedWriter` can be restarted.
+class BufferedWriter {
+ public:
+  using WriteFunction = std::function<void(grpc::ByteBuffer&&)>;
+
+  explicit BufferedWriter(WriteFunction&& write_func);
+
+  void Start();
+  void Stop();
+
+  bool empty() const {
+    return buffer_.empty();
+  }
+
+  bool IsStarted() const {
+    return is_started_;
+  }
+
+  void Enqueue(grpc::ByteBuffer&& bytes);
+  void OnSuccessfulWrite();
+
+  // Clears (but doesn't stop) the buffer. If there is an operation in progress,
+  // `OnSuccessfulWrite` must still be called for it.
+  void Clear();
```

Comment:
This API uses confusingly unrelated names:

  1. Enqueue (a buffer)
  2. OnSuccessfulWrite (about a buffer)
  3. Clear (when)?

These names aren't consistent in how they describe their actions: Enqueue seems to be about queueing but OnSuccessfulWrite expresses some higher-level concern. It's surprising there's no OnFailedWrite.

Clear uses naming that makes this more of a container like Enqueue, but has has semantics that aren't simple: If I'm required to call OnSuccessfulWrite and I call that before Clear isn't that going to start the next write? If I have to call OnSuccessfulWrite after Clear, what use case am I satisfying with the API? Just guessing, but it seems like I'd be calling Clear when I want to discard all pending writes, which I'd only do if I were throwing this object away.

Similarly, Enqueue doesn't promise that it's actually going to write anything. How does the caller know to call OnSuccessfulWrite? Does that write even have to succeed?

I think this API would be clearer if it got more opinionated more closely modeled our interaction with gRPC and the lifecycle of a streaming RPC.

So some concrete suggestions:
  * Rename OnSuccessfulWrite to something like DequeueNext() (to match the enqueue verb)
  * Remove or tighten up the meaning of Clear(). My guess is we'd only use this in the case of failure or shutdown. If so, rename to suit and don't preserve the started state after. If not used for those cases please add comments about when in the stream lifecycle it's appropriate to use this method.

Reply:
Regarding naming, the API (and implementation) still mixes bare queueing names and "write" concepts. We have Enqueue and DequeueNext vs DiscardUnstartedWrites and TryWrite. More thoughts:
  * Could `TryWrite` be renamed `TryDequeue`?
  * "Discard unstarted" definitely more clearly conveys what's going on than "clear" did.
  * The trailing "writes" in `DiscardUnstartedWrites` seems like it's either redundant (because this is a `BufferedWriter`) or an indication that we should carry that object (in the grammar sense) into other entry points (e.g. `EnqueueWrite`). You can make the case that the object of the function could be implicit here because the only thing this class does is manage writes. Then again, when reading code calling code like this:
    ```c++
    buffer_.DiscardUnstarted();
    ```
    ... it's not entirely obvious what's going on so I can see keeping the "writes" on the end too.

Ultimately up to you, but I'm arguing for consistency.

Regarding functionality and testing: I think it's important for classes to have a well defined role in the system. It's more important to be able to clearly describe how pieces fit together without messy caveats than to be able to cleanly test everything. Testability is also important but not at the expense of a clear role.

So, if managing the AsyncReaderWriter interaction is something that could logically make sense here, I'd prefer we introduce a testing seam that allows us to give this code the proper role than artificially bifurcate the responsibility just so we have an easy way to test.

A concrete way to achieve this would be to add a local interface describing our interaction with AsyncReaderWriter (perhaps even mirroring the AsyncReaderWriter API with virtual methods as the digital goods team did). From an efficiency point of view this is no worse than `std::function` and could be better if the compiler can discover that the virtual function is actually monomorphic.

I'm not arguing strongly one way or the other in terms of the role of the class--this is something that will only make sense in the context of the larger gRPC implementation.

Reply:
Thanks! Some thoughts:
* I tried adding `Write` to method names, so now there's `EnqueueWrite`, `DequeueNextWrite` (or should it be `DequeueWriteNext`?), `TryStartWrite` and `DiscardUnstartedWrites`.
* the renaming could go in reverse direction -- removing references to the queue from method names, something like `BufferedWrite` and `FinishWrite`/`NextWrite`/`GoToNextWrite`.
* I thought about how to make this class a little more self-sufficient. The best I could come up with was storing write operations, not byte buffers from which operations are created. In this case, the owner of this class supplies write operations for (potentially delayed) execution. This eliminates the need for `WriteFunction` callback but still allows testing (by mocking the operation). It's still the owner of `BufferedWriter`, not the writer itself, that receives notifications that writes are complete (and must call `Dequeue` on the writer). I think I like this more than the previous version, let me know what you think.
* apparently there's a mocking framework within googletest that allows [mocking non-virtual functions](https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#mocking-nonvirtual-methods). I don't know how feasible it is to use, but it sounds like it could be a better approach than creating wrappers for GRPC classes.